chain/client/src/stateless_validation/partial_witness/partial_witness_actor.rs

```diff
@@ -454,3 +467,12 @@ fn compress_witness(witness: &ChunkStateWitness) -> Result<EncodedChunkStateWitn
     );
     Ok(witness_bytes)
 }
+
+fn contracts_cache_contains_contract(
+    cache: &dyn ContractRuntimeCache,
+    contract_hash: &CodeHash,
+    runtime_config: &RuntimeConfig,
+) -> bool {
+    let cache_key = get_contract_cache_key(contract_hash.0, &runtime_config.wasm_config);
+    cache.memory_cache().contains(cache_key) || cache.has(&cache_key).is_ok_and(|has| has)
```

Comment:
Contract cache is not currently temporally consistent. That is, it is perfectly valid for either of the caches here to return `true` and then by the time it is time to execute contracts -- to return `false`. The lower level cache "evictions" don't also imply higher level cache evictions: it is possible for the `memory_cache` here to return `true`, but for the underlying to return `false` and then later for `memory_cache` to evict the contained contract through its natural FIFO mechanisms.

In the current implementation the most plausible way this might happen is if somebody ran a `rm -rf ~/.near/data/contracts/*` -- a really useful operation in certain scenarios -- between the check here and when the contract runtime accesses the contracts.

For the contract runtime this so far has been fine as a single cache access **internally** *is* atomic (i.e. `rm -rf` *while* executing a `cache.try_lookup` will deterministically either return an open file that you can still use to read its contents, even if it is no longer a named file, or it will fail...)

Now I wouldn't mind revoking rights for operators to do this while the node is running, but we may need to add some asserts around the area? In particular I don't see any changes to the code that accesses contracts from the "store" in this PR, but I imagine that ought to have an `expect("don't delete '~/.near/data/contracts/*' while node is running please")` in case the cache accesses end up turning up nothing and the contract hasn't been obtained here for whatever reason.

Reply:
Thanks for the elaborate comment! For the use case of witness validation those corner cases will result in a failure to endorse this particular witness, which could result in a missing chunk, but the node will continue operating without any issues and will be able to validate future witnesses.